File: challenges/xlang/cs/binarytrees.cs
using System;
using System.Text;

sealed class Tree { public Tree L, R; public Tree(Tree l, Tree r) { L = l; R = r; } }

static class BinaryTrees
{
    static Tree Make(int depth) => depth == 0 ? null : new Tree(Make(depth - 1), Make(depth - 1));
    static long Check(Tree t) => t == null ? 1 : 1 + Check(t.L) + Check(t.R);

    static int Main(string[] args)
    {
        int n = args.Length > 0 && int.TryParse(args[0], out int x) ? x : 10;
        int minDepth = 4;
        int maxDepth = minDepth + 2 > n ? minDepth + 2 : n;
        int stretchDepth = maxDepth + 1;
        Tree longLived = Make(maxDepth);
        StringBuilder outp = new StringBuilder();
        outp.Append($"stretch tree of depth {stretchDepth}\t check: {Check(Make(stretchDepth))}\n");
        for (int depth = minDepth; depth <= maxDepth; depth += 2)
        {
            long iterations = 1L << (maxDepth - depth + minDepth);
            long sum = 0;
            for (long i = 0; i < iterations; i++) sum += Check(Make(depth));
            outp.Append($"{iterations}\t trees of depth {depth}\t check: {sum}\n");
        }
        outp.Append($"long lived tree of depth {maxDepth}\t check: {Check(longLived)}\n");
        Console.Out.Write(outp.ToString());
        return 0;
    }
}

File: challenges/xlang/go/binarytrees.go
package main

import ("os"; "strconv"; "strings"; "fmt")

type Tree struct{ l, r *Tree }

func make_(depth int) *Tree {
	if depth == 0 { return nil }
	return &Tree{make_(depth - 1), make_(depth - 1)}
}
func check(t *Tree) int64 {
	if t == nil { return 1 }
	return 1 + check(t.l) + check(t.r)
}

func main() {
	n := 10
	if len(os.Args) > 1 { if x, err := strconv.Atoi(os.Args[1]); err == nil { n = x } }
	minDepth := 4
	maxDepth := n
	if minDepth+2 > n { maxDepth = minDepth + 2 }
	stretchDepth := maxDepth + 1
	longLived := make_(maxDepth)
	var out strings.Builder
	fmt.Fprintf(&out, "stretch tree of depth %d\t check: %d\n", stretchDepth, check(make_(stretchDepth)))
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := int64(1) << uint(maxDepth-depth+minDepth)
		var sum int64
		for i := int64(0); i < iterations; i++ { sum += check(make_(depth)) }
		fmt.Fprintf(&out, "%d\t trees of depth %d\t check: %d\n", iterations, depth, sum)
	}
	fmt.Fprintf(&out, "long lived tree of depth %d\t check: %d\n", maxDepth, check(longLived))
	os.Stdout.WriteString(out.String())
}
